src/rgw/rgw_sync_module_cloud.cc

```diff
@@ -0,0 +1,581 @@
+        return set_cr_error(retcode);
+      }
+      ldout(sync_env->cct, 20) << "info of remote obj: z=" << sync_env->source_zone
+        << " b=" << bucket_info.bucket << " k=" << key << dendl;
+      return set_cr_done();
+    }
+    return 0;
+  }
+
+};
+
+class RGWCloudRemove : public RGWAsyncRadosRequest {
+  CephContext *cct;
+  RGWBucketInfo bucket_info;
+
+  rgw_obj_key key;
+  RGWCloudAccess *cloud_access;
+protected:
+  int _send_request(){
+    int ret = cloud_access->remove_obj(bucket_info.bucket.name, key.name);
+    return ret;
+  }
+public:
+  RGWCloudRemove(RGWCoroutine *caller, RGWAioCompletionNotifier *cn,
+                       RGWBucketInfo& _bucket_info,
+                       const rgw_obj_key& _key,
+                       CephContext *_cct,
+                       RGWCloudAccess *_cloud_access) 
+                         : RGWAsyncRadosRequest(caller, cn),
+                           cct(_cct),
+                           bucket_info(_bucket_info),
+                           key(_key),
+                           cloud_access(_cloud_access)
+  { }
+};
+
+class RGWCloudRemoveCR : public RGWSimpleCoroutine {
+  CephContext *cct;
+  RGWBucketInfo bucket_info;
+
+  rgw_obj_key key;
+
+  RGWDataSyncEnv *sync_env;
+  RGWCloudRemove *req;
+  std::shared_ptr<RGWCloudAccess> cloud_access;
+public:
+  RGWCloudRemoveCR(RGWDataSyncEnv *_sync_env,
+                    RGWBucketInfo& _bucket_info,
+                    const rgw_obj_key& _key,
+                    std::shared_ptr<RGWCloudAccess>& _cloud_access) 
+                      : RGWSimpleCoroutine(_sync_env->store->ctx()), cct(_sync_env->store->ctx()),
+                        bucket_info(_bucket_info),
+                        key(_key),
+                        sync_env(_sync_env),
+                        req(NULL),
+                        cloud_access(_cloud_access) {
+
+  }
+  ~RGWCloudRemoveCR() {
+    request_cleanup();
+  }
+
+  void request_cleanup() {
+    if (req) {
+      req->finish();
+      req = NULL;
+    }
+  }
+
+  int send_request() {
+    req = new RGWCloudRemove(this, stack->create_completion_notifier(), bucket_info, key, cct, cloud_access.get());
+    sync_env->async_rados->queue(req);
+    return 0;
+  }
+
+  int request_complete() {
+    return req->get_ret_status();
+  }
+};
+
+class RGWCloudDataSyncModule : public RGWDataSyncModule {
+  std::string prefix;
+  std::map<std::string, std::string, ltstr_nocase>& config;
+  set<std::string> bucket_filter;
+  RGWCloudInfo cloud_info;
+  bool valid_config;
+  
+  bool check_bucket_filter(std::string& bucket_name) {
+    if (bucket_filter.empty()) {
+      return true;
+    }
+      
+    auto iter = bucket_filter.find(bucket_name);
+    return iter != bucket_filter.end() ? true : false;
+  }
+  
+public:
+  RGWCloudDataSyncModule(const std::string& _prefix, std::map<std::string, std::string, ltstr_nocase>& _config) 
+                          : prefix(_prefix), config(_config), valid_config(true) {
```

Comment:
the default value ture or false doesn't matter. the constucter of sync module is the best location to prase config